### ISSUE TITLE
Python poetry caret requirements

### ISSUE THREAD

**Issue by ryanhiebert** [dependabot/dependabot-core#3212]
Sorry for the sparse report, I tried to get into the code and see if I could identify or prove my theory, but I wasn't successful. I'm opening it up here in hopes that there may be someone watching that is able to confirm or disconfirm my theory faster than I am.

I use Poetry, and I have been frustrated by the pyproject.toml being updated unnecessarily. I have dependencies like `^1.1.0`, and when it updates from `1.1.0` to `1.2.0` it is updating the pyproject.toml file as well, which it shouldn't have to. The other day I noticed that it didn't update the pyproject.toml file on one dependency like that (of three that day), and it confused me, though I was happy for it.

My theory is that this requirement is being translated internally to work like `~1.1.0`, which notably *would* need to update the `pyproject.toml`, not just the lock file, in order to upgrade to `1.2.0`.

The correct behavior of Both caret dependencies and tilde dependencies are documented here: https://python-poetry.org/docs/versions/#caret-requirements.

Does this sound plausible? Is there any more guidance you might be able to give me in confirming this theory?

**Comment by Spectre5** [comment]
Did you figure out anything more on this?  I haven't actually used dependabot yet, but was looking into it.  I'd like it to update my Poetry lock file so that my tests will be run on new versions, but I do *NOT* want it to update the requirement in my `pyproject.toml` file.  For an application it is is fine to require later versions, but for a library you don't want to require the newest versions for all libraries or you'll make your library quickly incompatible with other tools.  A library should aim to keep the requirements as open as it can to stay compatible with any other libraries that another project may use.

That said, I'd want this to be optional.  For the --dev dependencies, it often would be great to keep those tools generally updated, even in the `pyproject.toml` file.

**Comment by ryanhiebert** [comment]
Thanks for asking! I currently do not have further update, either in confirmation or disconfirmation. I would appreciate it if we could do some experiments or sift through the code to confirm it.

**Comment by ryanhiebert** [comment]
If anyone at GitHub Actions or other CI is watching, fixing this bug could save quite a bit of resource utilization from automatically created pull requests having to be rebased, and the tests re-run.

**Comment by ryanhiebert** [comment]
Well, I don't think my hypothesis of the reason is likely to be correct, unfortunately. I have a dependabot PR that upgrades boto3 `^1.17.37` from `1.17.37` to `1.17.46` and changes the `pyproject.toml` file entry to `^1.17.46`, even though `1.17.46` is compatible with both `^1.17.37` and `~1.17.37`.

**Comment by ryanhiebert** [comment]
@Spectre5 : re-reading your question to me, I see that your concern is about versions being updated too much. I don't think there's any issue with that. If you set it to only upgrade major versions, I believe that's all dependabot will upgrade.

This is about having the pyproject.toml change, even when the minor version being updated is technically compatible with what's in the file already. I *do* want it to update my pyproject.toml, I just don't want it to do it for every pull request, because that makes other Dependabot pull requests conflict because it changes the hash that the lockfile is associated with. That forces dependabot to rebase the pull request, and then my tests have to run, and subsequent merges of other dependabot pull requests that could otherwise be done immediately have to wait for those to complete, thus significantly impeding the benefit I use dependabot for: saving time updating dependencies.

**Comment by ryanhiebert** [comment]
For those that are able, using `*` as the version specifier is an easy way to avoid this issue. For those that are not for whatever reason, it does appear that using `~` works correctly, and is able to give most of the same semantics that `^` is. Be sure to look up how the compatible release specifier works in [PEP 440](https://www.python.org/dev/peps/pep-0440/#compatible-release), because it's not _quite_ the same as the `^` in Poetry. Also, be aware that Poetry documents [tilde requirements](https://python-poetry.org/docs/versions/#tilde-requirements) as having different semantics than PEP 440 compatible releases, and it's not clear to me if that's a bug in the docs or arguably a bug in Poetry.

**Comment by ITProKyle** [comment]
I can confirm the same behaviour as described in the original post.

Example entry from `pyproject.toml`.

```toml
[tool.poetry.dependencies]
boto3 = "^1.16"
```

In the `poetry.lock` generated by my local install of poetry, I currently have it locked at `1.17.87`.

When dependabot runs, it is able to identify that there is a newer version available (1.17.90) but states `Requirements to unlock update_not_possible` and no PR is made. This would mean that dependabot is translating poetry's specification of `^1.16` to PEP 440's `~=1.16.0`. If it were translating it to PEP 440's `~=1.16` (the correct translation), it would be able to perform the update without widening the specification.

I have not looked through the code base to see how these are being translated so it's entirely possible the issue is on poetry's end if dependabot is exporting the dependencies when making these decisions.

FYI @ryanhiebert & @Spectre5, if you want dependabot to update your `poetry.lock` but not `pyproject.toml`, you only need to add `versioning-strategy: lockfile-only` to the dependabot configuration file. This will prevent it from widening the range of the specification in `pyproject.toml`.

```yaml
version: 2
updates:
  - directory: /
    open-pull-requests-limit: 5
    package-ecosystem: pip
    schedule:
      interval: daily
      time: "08:00"
      timezone: America/New_York
    versioning-strategy: lockfile-only
```

**Comment by deivid-rodriguez** [comment]
Hei! I think the root cause here is lack of support for the `increase-if-necessary` versioning strategy.

The good news is that it will be fixed by #5605! 🎉
